Flaky test failures: 'stream' != 'streamstream'

I'm seeing unexplainable test failures for cells with print statements, for example 
```
=================================== FAILURES ===================================
____________________________________ cell 8 ____________________________________
Notebook cell execution failed
Cell 8: Cell outputs differ

Input:
trace(system_test, 'unit_i', 10)

Traceback: mismatch 'output_type'
<<<<<<<<<<<< Reference output from ipynb file:
stream
============ disagrees with newly computed (test) output:
streamstream
>>>>>>>>>>>>
```
The cell in question runs a function that print()s a few lines then returns None
```
trace(system_test, 'unit_i', 10)
```
I have not been able to reproduce with a simple example, but you can see the failing notebook at
  https://github.com/fritzo/pomagma/blob/47eecf8/src/examples/definable_systems.ipynb
where I've `if 0` guarded the final cell to work around this issue.

Here's my system (running python 2.7)
```
$ pip freeze
altgraph==0.10.2
apipkg==1.4
appnope==0.1.0
appscript==1.0.1
argh==0.15.1
autoflake==0.6.6
autopep8==1.2.4
backports.shutil-get-terminal-size==1.0.0
backports.ssl-match-hostname==3.4.0.2
bdist-mpkg==0.5.0
BeautifulSoup==3.2.1
bleach==1.5.0
bonjour-py==0.3
boto==2.43.0
bottle==0.11.6
cdiff==0.9.6
certifi==2015.9.6.2
configparser==3.5.0
contextlib2==0.5.4
decorator==4.0.10
docformatter==0.8
entrypoints==0.2.2
enum34==1.1.6
execnet==1.4.1
flake8==3.2.1
funcsigs==1.0.2
functools32==3.2.3.post2
goftests==0.2.0
gprof2dot==2015.12.1
html5lib==0.9999999
hypothesis==3.6.1
ipykernel==4.5.2
ipython==5.1.0
ipython-genutils==0.1.0
ipywidgets==5.2.2
isodate==0.4.8
isort==4.2.5
Jinja2==2.8.1
jsonschema==2.5.1
jupyter==1.0.0
jupyter-client==4.4.0
jupyter-console==5.0.0
jupyter-core==4.2.1
jupyterthemes==0.13.9
lesscpy==0.12.0
logic==0.1.10
macholib==1.5.1
Markdown==2.2.0
MarkupSafe==0.23
matplotlib==1.3.1
mccabe==0.5.3
mistune==0.7.3
mock==2.0.0
modulegraph==0.10.4
multipledispatch==0.4.9
nbconvert==5.0.0
nbformat==4.2.0
nbval==0.5
nose==1.3.7
notebook==4.3.1
numpy==1.6.2
pandocfilters==1.4.1
parsable==0.2.2
pathlib2==2.1.0
pbr==1.10.0
pep8==1.7.0
pexpect==4.2.1
pickleshare==0.7.4
ply==3.9
-e git+git@github.com:fritzo/pomagma.git@47eecf89dda27a81f7e7900597c2ab5f0f8fbce3#egg=pomagma
prompt-toolkit==1.0.9
protobuf==2.6.1
psutil==4.4.2
ptyprocess==0.5.1
py==1.4.31
py2app==0.7.3
pycodestyle==2.2.0
pyflakes==1.3.0
pyformat==0.6
Pygments==1.5
pyobjc-core==2.5.1
pyobjc-framework-Accounts==2.5.1
pyobjc-framework-AddressBook==2.5.1
pyobjc-framework-AppleScriptKit==2.5.1
pyobjc-framework-AppleScriptObjC==2.5.1
pyobjc-framework-Automator==2.5.1
pyobjc-framework-CFNetwork==2.5.1
pyobjc-framework-Cocoa==2.5.1
pyobjc-framework-Collaboration==2.5.1
pyobjc-framework-CoreData==2.5.1
pyobjc-framework-CoreLocation==2.5.1
pyobjc-framework-CoreText==2.5.1
pyobjc-framework-DictionaryServices==2.5.1
pyobjc-framework-EventKit==2.5.1
pyobjc-framework-ExceptionHandling==2.5.1
pyobjc-framework-FSEvents==2.5.1
pyobjc-framework-InputMethodKit==2.5.1
pyobjc-framework-InstallerPlugins==2.5.1
pyobjc-framework-InstantMessage==2.5.1
pyobjc-framework-LatentSemanticMapping==2.5.1
pyobjc-framework-LaunchServices==2.5.1
pyobjc-framework-Message==2.5.1
pyobjc-framework-OpenDirectory==2.5.1
pyobjc-framework-PreferencePanes==2.5.1
pyobjc-framework-PubSub==2.5.1
pyobjc-framework-QTKit==2.5.1
pyobjc-framework-Quartz==2.5.1
pyobjc-framework-ScreenSaver==2.5.1
pyobjc-framework-ScriptingBridge==2.5.1
pyobjc-framework-SearchKit==2.5.1
pyobjc-framework-ServiceManagement==2.5.1
pyobjc-framework-Social==2.5.1
pyobjc-framework-SyncServices==2.5.1
pyobjc-framework-SystemConfiguration==2.5.1
pyobjc-framework-WebKit==2.5.1
pyOpenSSL==0.13.1
pyparsing==2.0.1
PySMT==0.6.1
pytest==3.0.3
pytest-profiling==1.1.1
pytest-timeout==1.0.0
pytest-xdist==1.15.0
python-dateutil==1.5
python-magic==0.4.6
pytz==2013.7
PyYAML==3.10
pyzmq==16.0.0
qtconsole==4.2.1
requests==2.12.4
RunSnakeRun==2.0.4
s3cmd==1.5.2
scipy==0.13.0b1
selenium==2.33.0
simplegeneric==0.8.1
simplejson==3.10.0
six==1.10.0
snakeviz==0.4.0
splinter==0.5.4
SquareMap==1.0.4
stevedore==1.18.0
terminado==0.6
testpath==0.3
toolz==0.8.0
tornado==4.2.1
traitlets==4.3.1
unification==0.2.2
unify==0.2
untokenize==0.1.1
vboxapi==1.0
veritable===0.9.8preBUILD-NUMBER
virtualenv==15.0.3
virtualenv-clone==0.2.6
virtualenvwrapper==4.7.2
wcwidth==0.1.7
widgetsnbextension==1.2.6
xattr==0.6.4
yapf==0.14.0
zope.interface==4.1.1
```

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -196,8 +196,7 @@ class IPyNbFile(pytest.File):
             'traceback',
             #'text/latex',
             'prompt_number',
-            'stdout',
-            'stream',
+            'output_type',
             'name',
             'execution_count',
         )
@@ -349,85 +348,50 @@ class IPyNbCell(pytest.Item):
         # Use stored skips unless passed a specific value
         skip_compare = skip_compare or self.parent.skip_compare
 
-        # For every different key, we will store the outputs in a
-        # single string, in a dictionary with the same keys
-        # At the end, every dictionary entry will be compared
-        # We skip the unimportant keys in the 'skip_compare' list
-        #
-        # We concatenate the outputs because the ipython notebook produces
-        # them in a random number of dictionaries. So, it is easier
-        # to compare only one chunk of data
-        testing_outs = defaultdict(str)
-        reference_outs = defaultdict(str)
+        test = transform_streams_for_comparison(test)
+        ref = transform_streams_for_comparison(ref)
 
-        # Check the references (embedded notebook outputs)
-        # and start appendind the outputs for every
-        # different key. The entries of every output have the structure:
+        # We reformat outputs into a dictionaries where
+        # key:
+        #   - all keys on output except 'data' and those in skip_compare
+        #   - all keys on 'data' except those in skip_compare, i.e. data is flattened
+        # value:
+        #   - list of all corresponding values for that key, i.e. for all outputs
         #
-        # {'output_type': 'stream', 'stream': 'stdout',
-        #  'text': "The time is: 11:44:21\nToday's date is: 13/03/15\n"}
-        #
-        # We discard the keys from the skip_compare list
+        # This format allows to disregard the relative order of dissimilar
+        # output keys, while still caring about the order of those that share
+        # a key.
+        testing_outs = defaultdict(list)
+        reference_outs = defaultdict(list)
+
         for reference in ref:
             for key in reference.keys():
+                # We discard the keys from the skip_compare list:
                 if key not in skip_compare:
-
-                    # In the EXECUTION, we already processed the display_data
-                    # (or execute_count)
-                    # kind of dictionary entries. display_data has a 'data'
-                    # sub dictionary which contains the relevant information
-                    # about the Figure: 'text/plain', 'image/png', ...
-                    #
-                    # EXAMPLES:
-                    #
-                    # display_data type:
-                    # {'output_type': 'display_data', 'image/png': 'iVBORw0...
-                    #  'text/plain': <matplotlib.figure.Figure at 0x7f9ca97cc890>
-                    #  'metadata': {} }
-                    #
-                    #
-                    # Hence, we look into these sub dictionary entries and
-                    # append them to the corresponding dictionary entry
-                    # in the reference outputs
-                    #
+                    # Flatten out MIME types from data of display_data and execute_result
                     if key == 'data':
                         for data_key in reference[key].keys():
-                            # Filter the keys in the SUB-dictionary again
+                            # Filter the keys in the SUB-dictionary again:
                             if data_key not in skip_compare:
-                                reference_outs[data_key] += self.sanitize(reference[key][data_key])
-
-
-                    # NOTICE: that execute_result (similar for figures than
-                    # display_data but without the png or picture hex)
-                    # has an 'execution_count' key
-                    # which we skip because is not relevant for now.
-                    # We could use this in the future if we wanted executions
-                    # in the same order than the reference
-                    #
-                    # execute_result type:
-                    # {'output_type': 'execute_result', 'execution_count': 9,
-                    #  'text/plain': '<matplotlib.image.AxesImage at 0x7f9ca8f058d0>',
-                    #  'metadata': {}}
+                                reference_outs[data_key].append(self.sanitize(reference[key][data_key]))
 
                     # Otherwise, just create a normal dictionary entry from
                     # one of the keys of the dictionary
                     else:
                         # Create the dictionary entries on the fly, from the
                         # existing ones to be compared
-                        reference_outs[key] += self.sanitize(reference[key])
+                        reference_outs[key].append(self.sanitize(reference[key]))
 
         # the same for the testing outputs (the cells that are being executed)
-        # display_data cells were already processed! (see the execution loop)
         for testing in test:
             for key in testing.keys():
                 if key not in skip_compare:
                     if key == 'data':
                         for data_key in testing[key].keys():
-                            # Filter the keys in the SUB-dictionary again
                             if data_key not in skip_compare:
-                                testing_outs[data_key] += self.sanitize(testing[key][data_key])
+                                testing_outs[data_key].append(self.sanitize(testing[key][data_key]))
                     else:
-                        testing_outs[key] += self.sanitize(testing[key])
+                        testing_outs[key].append(self.sanitize(testing[key]))
 
         # The traceback from the comparison will be stored here.
         self.comparison_traceback = []
@@ -443,32 +407,55 @@ class IPyNbCell(pytest.Item):
                     + bcolors.ENDC)
                 return False
 
-            # Compare the large string from the corresponding dictionary entry
+            # Get output values for dictionary entries.
             # We use str() to be sure that the unicode key strings from the
-            # reference are also read from the testing dictionary
-            if testing_outs[str(key)] != reference_outs[key]:
-
-                # print testing_outs[key]
-                # print reference_outs[key]
-
+            # reference are also read from the testing dictionary:
+            test_values = testing_outs[str(key)]
+            ref_values = reference_outs[key]
+            if len(test_values) != len(ref_values):
+                # The number of outputs for a specific MIME type differs
                 self.comparison_traceback.append(
                     bcolors.OKBLUE
-                    + " mismatch '%s'\n" % key
+                    + 'dissimilar number of outputs for key "%s"' % key
                     + bcolors.FAIL
-                    + "<<<<<<<<<<<< Reference output from ipynb file:"
-                    + bcolors.ENDC)
-                self.comparison_traceback.append(_trim_base64(reference_outs[key]))
+                    + "<<<<<<<<<<<< Reference outputs from ipynb file:"
+                    + bcolors.ENDC
+                )
+                for val in ref_values:
+                    self.comparison_traceback.append(_trim_base64(val))
                 self.comparison_traceback.append(
                     bcolors.FAIL
-                    + '============ disagrees with newly computed (test) output:  '
+                    + '============ disagrees with newly computed (test) output:'
                     + bcolors.ENDC)
-                self.comparison_traceback.append(_trim_base64(testing_outs[str(key)]))
+                for val in test_values:
+                    self.comparison_traceback.append(_trim_base64(val))
                 self.comparison_traceback.append(
                     bcolors.FAIL
                     + '>>>>>>>>>>>>'
                     + bcolors.ENDC)
-
                 return False
+
+            for test_out, ref_out in zip(test_values, ref_values):
+                # Compare the individual values
+                if test_out != ref_out:
+                    self.comparison_traceback.append(
+                        bcolors.OKBLUE
+                        + " mismatch '%s'\n" % key
+                        + bcolors.FAIL
+                        + "<<<<<<<<<<<< Reference output from ipynb file:"
+                        + bcolors.ENDC)
+                    self.comparison_traceback.append(_trim_base64(ref_out))
+                    self.comparison_traceback.append(
+                        bcolors.FAIL
+                        + '============ disagrees with newly computed (test) output:'
+                        + bcolors.ENDC)
+                    self.comparison_traceback.append(_trim_base64(test_out))
+                    self.comparison_traceback.append(
+                        bcolors.FAIL
+                        + '>>>>>>>>>>>>'
+                        + bcolors.ENDC)
+
+                    return False
         return True
 
 
@@ -653,17 +640,7 @@ class IPyNbCell(pytest.Item):
             # to obtain the 'text' and 'image/png' information
             elif msg_type in ('display_data', 'execute_result'):
                 out['metadata'] = reply['metadata']
-                out['data'] = {}
-                for mime, data in six.iteritems(reply['data']):
-                    # This could be useful for reference or backward compatibility
-                    #     attr = mime.split('/')[-1].lower()
-                    #     attr = attr.replace('+xml', '').replace('plain', 'text')
-                    #     setattr(out, attr, data)
-
-                    # Return the relevant entries from data:
-                    # plain/text, image/png, execution_count, etc
-                    # We could use a mime types list for this (MAYBE)
-                    out.data[mime] = data
+                out['data'] = reply['data']
                 outs.append(out)
 
                 if msg_type == 'execute_result':
@@ -672,7 +649,7 @@ class IPyNbCell(pytest.Item):
 
             # if the message is a stream then we store the output
             elif msg_type == 'stream':
-                out.stream = reply['name']
+                out.name = reply['name']
                 out.text = reply['text']
                 outs.append(out)
 
@@ -701,6 +678,8 @@ class IPyNbCell(pytest.Item):
             else:
                 print("unhandled iopub msg:", msg_type)
 
+        outs[:] = coalesce_streams(outs)
+
         # Compare if the outputs have the same number of lines
         # and throw an error if it fails
         # if len(outs) != len(self.cell.outputs):
@@ -708,7 +687,7 @@ class IPyNbCell(pytest.Item):
         #     failed = True
         failed = False
         if self.options['check']:
-            if not self.compare_outputs(outs, self.cell.outputs):
+            if not self.compare_outputs(outs, coalesce_streams(self.cell.outputs)):
                 failed = True
 
         # If the comparison failed then we raise an exception.
@@ -726,8 +705,6 @@ class IPyNbCell(pytest.Item):
                                                        'traceback',
                                                        'text/latex',
                                                        'prompt_number',
-                                                       'stdout',
-                                                       'stream',
                                                        'output_type',
                                                        'name',
                                                        'execution_count'
@@ -772,6 +749,64 @@ class IPyNbCell(pytest.Item):
         for regex, replace in six.iteritems(self.parent.sanitize_patterns):
             s = re.sub(regex, replace, s)
         return s
+
+
+carriagereturn_pat = re.compile(r'.*\r(?=[^\n])')
+backspace_pat = re.compile(r'[^\n]\b')
+
+
+def coalesce_streams(outputs):
+    """
+    Merge all stream outputs with shared names into single streams
+    to ensure deterministic outputs.
+
+    Parameters
+    ----------
+    outputs : iterable of NotebookNodes
+        Outputs being processed
+    """
+    if not outputs:
+        return outputs
+
+    new_outputs = []
+    streams = {}
+    for output in outputs:
+        if (output.output_type == 'stream'):
+            if output.name in streams:
+                streams[output.name].text += output.text
+            else:
+                new_outputs.append(output)
+                streams[output.name] = output
+        else:
+            new_outputs.append(output)
+
+    # process \r and \b characters
+    for output in streams.values():
+        backspace_pat
+        old = output.text
+        while len(output.text) < len(old):
+            old = output.text
+            # Cancel out anything-but-newline followed by backspace
+            output.text = backspace_pat.sub('', output.text)
+        # Replace all carriage returns not followed by newline
+        output.text = carriagereturn_pat.sub('', output.text)
+
+    return new_outputs
+
+
+def transform_streams_for_comparison(outputs):
+    """Makes failure output for streams better by having key be the stream name"""
+    new_outputs = []
+    for output in outputs:
+        if (output.output_type == 'stream'):
+            # Transform output
+            new_outputs.append({
+                'output_type': 'stream',
+                output.name: output.text,
+            })
+        else:
+            new_outputs.append(output)
+    return new_outputs
 
 
 def get_sanitize_patterns(string):

--- a/tests/test_coalesce.ipynb
+++ b/tests/test_coalesce.ipynb
@@ -1,0 +1,178 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from __future__ import print_function\n",
+    "import sys"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Foo!\n",
+      "Bar!\n"
+     ]
+    }
+   ],
+   "source": [
+    "print('Foo!')\n",
+    "print('Bar!')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Bar!\n",
+      "Foo!\n"
+     ]
+    }
+   ],
+   "source": [
+    "print('Bar!', file=sys.stderr)\n",
+    "print('Foo!', file=sys.stderr)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Combo breaker!\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Bar!\n",
+      "Foo!\n"
+     ]
+    }
+   ],
+   "source": [
+    "print('Bar!', file=sys.stderr)\n",
+    "print('Combo breaker!', file=sys.stdout)\n",
+    "print('Foo!', file=sys.stderr)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Bar!\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Combo breaker!\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Foo!\n"
+     ]
+    }
+   ],
+   "source": [
+    "print('Bar!', file=sys.stderr)\n",
+    "sys.stderr.flush()\n",
+    "print('Combo breaker!', file=sys.stdout)\n",
+    "sys.stdout.flush()\n",
+    "print('Foo!', file=sys.stderr)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Bar!\n",
+      "Foo!\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Combo breaker!\n"
+     ]
+    }
+   ],
+   "source": [
+    "import time\n",
+    "import numpy as np\n",
+    "print('Bar!', file=sys.stderr)\n",
+    "# Do not run the next line when regenerating test!\n",
+    "sys.stderr.flush()\n",
+    "print('Combo breaker!', file=sys.stdout)\n",
+    "sys.stderr.flush()\n",
+    "print('Foo!', file=sys.stderr)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": []
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "Python 3",
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.5.2"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}

